Include more diagnostic updates for next Dart 3 beta (#4738)
dart-lang/site-www: src/tools/diagnostic-messages.md

}
{% endprettify %}

### non_exhaustive_switch_expression

_The type '{0}' is not exhaustively matched by the switch cases since it doesn't
match '{1}'._

#### Description

The analyzer produces this diagnostic when a `switch` expression is
missing a case for one or more of the possible values that could flow
through it.

#### Example

The following code produces this diagnostic because the switch expression
doesn't have a case for the value `E.three`:

{% prettify dart tag=pre+code %}
enum E { one, two, three }

String f(E e) => [!switch!] (e) {
    E.one => 'one',
    E.two => 'two',
  };
{% endprettify %}

#### Common fixes

> Add a case for each of the constants that aren't currently being matched:

Reviewer: **issue:** Active voice. Condition, then action.
```suggestion
For each constant going unmatched, add a case:
```

{% prettify dart tag=pre+code %}
enum E { one, two, three }

String f(E e) => switch (e) {
    E.one => 'one',
    E.two => 'two',
    E.three => 'three',
  };
{% endprettify %}

> If the missing values don't need to be matched, then add  a wildcard
> pattern:

Reviewer: **suggestion:** Tighten this up some. Also, semantic break.
```suggestion
If you don't need an exact match for the missing values,
add a wildcard pattern:
```

{% prettify dart tag=pre+code %}
enum E { one, two, three }

String f(E e) => switch (e) {
    E.one => 'one',
    E.two => 'two',
    _ => 'unknown',
  };
{% endprettify %}

> But be aware that adding a wildcard pattern will cause any future values
> of the type to also be handled, so you will have lost the ability for the
> compiler to warn you if the `switch` needs to be updated.

Reviewer: **issue:** Sentence is too long and lacks readability.
```suggestion
If you add a wildcard pattern, it handles all future values for the type.
The compiler can't warn you when you need to update the `switch`.
```

}
{% endprettify %}

### positional_field_in_object_pattern

_Object patterns can only use named fields._

#### Description

> The analyzer produces this diagnostic when an object pattern contains a
> field that doesn't have a getter name. The fields provide a pattern to
> match against the value returned by a getter, and not specifying the name
> of the getter means that there's no way to access the value that the
> pattern is intended to match against.

Reviewer: **issue:** Second sentence is too long (39 words) and lacks readability. (Grade 13)
```suggestion
The analyzer returns this message when an object pattern contains
a field without a getter name. The fields provide a pattern to match.
This pattern matches against the value that a getter returns.
 If you don't provide the name of the getter,
 you can't access the value that the pattern matches.
```

This change drops readability to Grade 7.

}
{% endprettify %}

### rest_element_in_map_pattern

_A map pattern can't contain a rest pattern._

#### Description

> The analyzer produces this diagnostic when a map pattern contains a rest
> pattern. The matching for map patterns already allows the map to have
> more keys than those explicitly given in the pattern, so a rest pattern
> wouldn't add anything.

Reviewer: **issue:** Readability.
```suggestion
The analyzer returns this message when a map pattern contains a rest pattern.
Map pattern matching allows the map to have more keys than those given in the pattern.
A rest pattern adds no benefit.
```
